VERSIONING.md

```diff
@@ -0,0 +1,220 @@
+          `go.mod` files (e.g., `module go.opentelemetry.io/collector/v2`, `require
+          go.opentelemetry.io/collector/v2 v2.0.1`) and in the package import path
+          (e.g., `import "go.opentelemetry.io/collector/v2/component"`). This includes the
+          paths used in `go get` commands (e.g., `go get
+          go.opentelemetry.io/collector/v2@v2.0.1`.  Note there is both a `/v2` and a
+          `@v2.0.1` in that example. One way to think about it is that the module
+          name now includes the `/v2`, so include `/v2` whenever you are using the
+          module name).
+        * If a module is version `v0` or `v1`, do not include the major version in
+          either the module path or the import path.
+    * Modules will be used to encapsulate signals and components.
+        * Experimental modules still under active development will be versioned at
+          `v0` to imply the stability guarantee defined by
+          [semver](https://semver.org/spec/v2.0.0.html#spec-item-4).
+
+          > Major version zero (0.y.z) is for initial development. Anything MAY
+          > change at any time. The public API SHOULD NOT be considered stable.
+
+        * Mature modules for which we guarantee a stable public API will be versioned
+          with a major version greater than `v0`.
+            * The decision to make a module stable will be made on a case-by-case
+              basis by the maintainers of this project.
+        * Experimental modules will start their versioning at `v0.0.0` and will
+          increment their minor version when backwards incompatible changes are
+          released and increment their patch version when backwards compatible
+          changes are released.
+        * All stable modules that use the same major version number will use the
+          same entire version number.
+            * Stable modules may be released with an incremented minor or patch
+              version even though that module has not been changed, but rather so
+              that it will remain at the same version as other stable modules that
+              did undergo change.
+            * When an experimental module becomes stable a new stable module version
+              will be released and will include this now stable module. The new
+              stable module version will be an increment of the minor version number
+              and will be applied to all existing stable modules as well as the newly
+              stable module being released.
+* Versioning of the associated [contrib
+  repository](https://github.com/open-telemetry/opentelemetry-collector-contrib) of
+  this project will be idiomatic of a Go project using [Go
+  modules](https://github.com/golang/go/wiki/Modules).
+    * [Semantic import
+      versioning](https://github.com/golang/go/wiki/Modules#semantic-import-versioning)
+      will be used.
+        * Versions will comply with [semver 2.0](https://semver.org/spec/v2.0.0.html).
```

Comment:
Why does contrib have no "New methods may be added to exported API interfaces" exception?

Reply:
That exception was put in place to allow for the OTel spec to grow through accretion and with the knowledge that, at least in the Go client library, API and SDK changes would be made in lockstep due to sharing a repository and CI system.  We also expected that the number of implementations of those interfaces would be small and SDK authors would need to accept the responsibility of keeping up-to-date.  In practice, this turned out to be a very small number of interfaces: `TracerProvider`, `Tracer`, `Span`, and, by extension, the SDK's `ReadOnlySpan` and `ReadWriteSpan`.

I'm not even 100% sure it's appropriate for the collector core, the interfaces there are not defined in the specification and are intended to be broadly implemented.  I think it's worth having this discussion, though, about whether there are any such interfaces in the collector.

As for collector-contrib, I would expect even fewer outward-facing interfaces there but for those to be even more important to keep stable as they would likely be targeted for end-user implementation rather than component developer implementation.

Reply:
👍  that makes sense, let's continue the discussion for the need of these on https://github.com/open-telemetry/opentelemetry-collector/pull/3476#discussion_r656279054